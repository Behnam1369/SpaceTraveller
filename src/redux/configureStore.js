import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import missionReducer from './missions';
import { rocketsReducer } from './rockets';

const rootReducer = combineReducers({
  rockets: rocketsReducer,
  missions: missionReducer,
});

const store = createStore(rootReducer, applyMiddleware(thunk));

export default store;
