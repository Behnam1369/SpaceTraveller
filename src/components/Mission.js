import { React, useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { joinMission, leaveMission } from '../redux/missions';

function Mission(props) {
  const {
    id, name, description, joined,
  } = props;
  const [expanded, setExpanded] = useState(false);
  const dispatch = useDispatch();

  const handleReadMore = () => {
    setExpanded(!expanded);
  };

  const handleContribution = () => {
    if (!joined) {
      dispatch(joinMission(id));
    } else {
      dispatch(leaveMission(id));
    }
  };

  // description.split(' ').filter((el, i) => i < 50).join(' ')

  return (
    <tr className="mission">
      <td>{name}</td>
      {
        expanded || description.split(' ').length < 70
          ? (
            <td>
              {' '}
              {description}
              {' '}
            </td>
          )
          : (
            <td>
              {`${description.split(' ').filter((el, i) => i < 50).join(' ')} ...`}
              <button type="button" onClick={handleReadMore}>Read more</button>
            </td>
          )
      }
      <td>
        <span className={joined ? 'joined' : ''}>
          {' '}
          {joined ? 'Active Member' : 'NOT A MEMBER'}
          {' '}
        </span>
      </td>
      <td>
        <button type="button" className={joined ? 'joined' : ''} onClick={handleContribution}>
          {joined ? 'Leave Mission' : 'Join Mission'}
        </button>
      </td>
    </tr>
  );
}

Mission.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  joined: PropTypes.bool,
};

Mission.defaultProps = {
  joined: false,
};

export default Mission;
