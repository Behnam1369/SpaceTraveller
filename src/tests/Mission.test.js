import TestRenderer from 'react-test-renderer';
// import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import Mission from '../components/Mission';
import '@testing-library/jest-dom';
import configureStore from '../redux/configureStore';

describe('Testing Mission', () => {
  it('Testing Mission component snapshot', () => {
    const testRenderer = TestRenderer.create(
      <Provider store={configureStore}>
        <Mission
          key="key1"
          id="key1"
          description="A Sample Mission Description"
          name="Impossible Mission"
        />
      </Provider>,
    ).toJSON();
    expect(testRenderer).toMatchSnapshot();
  });
});
