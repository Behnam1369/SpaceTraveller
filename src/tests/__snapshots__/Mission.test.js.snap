// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Mission Testing Mission component snapshot 1`] = `
<tr
  className="mission"
>
  <td>
    Impossible Mission
  </td>
  <td>
     
    A Sample Mission Description
     
  </td>
  <td>
    <span
      className=""
      data-testid="badge"
    >
       
      NOT A MEMBER
       
    </span>
  </td>
  <td>
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Join Mission
    </button>
  </td>
</tr>
`;
